handlers/newObjects.go

```
@@ -305,3 +314,143 @@ func (a *RestAPI) NewHeadByAttribute(ctx echo.Context, containerID apiserver.Con
+	if btoken != nil {
+		attachBearer(&prmRange, btoken)
+	}
+
+	resObj, err := a.pool.ObjectRangeInit(ctx.Request().Context(), addr.Container(), addr.Object(), offset, length, a.signer, prmRange)
+	if err != nil {
+		return ctx.JSON(http.StatusBadRequest, util.NewErrorResponse(err))
+	}
+
+	payload := io.ReadCloser(resObj)
+
+	if len(contentType) == 0 {
+		// Determine the Content-Type from the payload head.
+		var payloadHead []byte
+
+		contentType, payloadHead, err = readContentType(length, func(uint64) (io.Reader, error) {
+			return payload, nil
+		})
+		if err != nil {
+			resp := a.logAndGetErrorResponse("invalid  ContentType", err)
+			return ctx.JSON(http.StatusBadRequest, resp)
+		}
+
+		// A piece of `payload` was read and is stored in `payloadHead`.
+		// RangeReader allows reading data from both `payloadHead` and `payload` starting from position `start`,
+		// regardless of where the `start` is.
+		a.log.Debug("RangeReader params",
+			zap.Int("payloadHead length", len(payloadHead)),
+			zap.Uint64("length", length),
+			zap.Uint64("start", start))
+		rangeReader := NewRangeReader(payload, payloadHead, length, start)
+		payload = readCloser{rangeReader, rangeReader}
+	}
+
+	ctx.Response().Header().Set("Content-Type", contentType)
+	ctx.Response().Header().Set(accessControlAllowOriginHeader, "*")
+
+	ctx.Response().Header().Set("Content-Range", fmt.Sprintf("bytes %d-%d/%d", start, end, payloadSize))
+	ctx.Response().Header().Set("Content-Length", strconv.FormatUint(end-start+1, 10))
+
+	return ctx.Stream(http.StatusPartialContent, contentType, payload)
+}
+
+func getRangeParams(rangeParam string, payloadSize uint64) (end, start uint64, err error) {
+	const prefix = "bytes="
+
+	// Preliminary checks.
+	if payloadSize == 0 {
+		return 0, 0, errors.New("zero payload size")
+	}
+	if !strings.HasPrefix(rangeParam, prefix) {
+		return 0, 0, errors.New("bytes= prefix required")
+	}
+	arr := strings.Split(strings.TrimPrefix(rangeParam, prefix), "-")
```

It's not that easy, you can have multipart range request (https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests#multipart_ranges).

---

But I doubt it's often used, so you can detect it and throw an "unsupported multipart range request" for now and create an issue for the future.

---

I've created https://github.com/nspcc-dev/neofs-rest-gw/issues/242